AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,886 @@
+                                                               assert(inNumberOfFrames <= kMaximumFramesPerBuffer);
+
+                                                               AudioBufferList *bufferList = context->bufferList;
+                                                               int8_t *audioBuffer = (int8_t *)bufferList->mBuffers[0].mData;
+                                                               UInt32 audioBufferSizeInBytes = bufferList->mBuffers[0].mDataByteSize;
+
+                                                               if (context->deviceContext) {
+                                                                   /*
+                                                                    * Pull decoded, mixed audio data from the media engine into the
+                                                                    * AudioUnit's AudioBufferList.
+                                                                    */
+                                                                   TVIAudioDeviceReadRenderData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+
+                                                               } else {
+
+                                                                   /*
+                                                                    * Return silence when we do not have the playout device context. This is the
+                                                                    * case when the remote participant has not published an audio track yet.
+                                                                    * Since the audio graph and audio engine has been setup, we can still play
+                                                                    * the music file using the AVAudioEngine.
+                                                                    */
+                                                                   memset(audioBuffer, 0, audioBufferSizeInBytes);
+                                                               }
+
+                                                               return bufferList;
+                                                           }];
+        if (!success) {
+            NSLog(@"Failed to set the manual rendering block");
+            return NO;
+        }
+
+        success = [_engine startAndReturnError:&error];
+        if (!success) {
+            NSLog(@"Failed to start AVAudioEngine, error = %@", error);
+            return NO;
+        }
+    }
+    return YES;
+}
+
+- (void)teardownAudioEngine {
+    [self teardownPlayer];
+    [_engine stop];
+    _engine = nil;
+}
+
+- (void)playMusic {
+    if (!_engine) {
+        NSLog(@"Cannot play music. AudioEngine has not been created yet.");
+        return;
+    }
+
+    if (_player.isPlaying) {
+        [_player stop];
+    }
+
+    /*
+     * We will attach an audio player node to AVAudioEnine to the main mixer.
```

Comment:
Grammar and spelling. `AVAudioEngine`.

My take - `Attach an AVAudioPlayerNode as an input to the main mixer.`